Implement non-fullscreen streaming rendering

quark-zju opened the pull request:
This matches `less -F -X` behavior before entering the full screen - lines are printed as the pager recieves them.

Demo: https://asciinema.org/a/KZbc9K1elnz9alqxbCq6k1XLK

I had a hard time naming the feature. Better names are welcome.

markbt commented:
This removes the delay fullscreen mode, which is my preferred mode of operation (and, in fact, my primary motivator for writing `sp`).  I want to use alternate screen mode if and only if the pager is necessary.  I'm happy to pay the ~2s delay of getting no output before the pager kicks in (in practice I find most things are either <1s or much longer).

I'm happy to add new modes, but please don't remove delay :)

quark-zju commented:
By using the alternate screen, do you mainly want the behavior of cleaning up the pager content on exit? If that's the case, how about:
- Still print content immediately before entering full screen.
- When entering full screen, erasing lines printed so it appears nothing was printed in the non-alternate-screen.
- Always use the alternate screen for the full screen mode.

markbt commented:
That's part of it.  I do like how it leaves the terminal output clean after the pager exits, so if we can emulate that, too, that'll be much better than less's current behaviour.

The other part is for commands that produce output slowly.  By jumping into the fullscreen pager after 2s, it's possible to see that something is happening, and also to do things like search or scroll through the limited amount of output that's already there.  It's quite nice to see the animated throbber in the bottom left, plus have the new content pop into view as it comes in.

The latter one is probably acceptable if we make any keypress jump into the fullscreen pager, too.  However, I have been using the delay mode for a year now, and I am quite fond of it's behaviour.

quark-zju commented:
I got a version with the delayed mode integerated. I will update this after cleaning it up.

quark-zju commented:
I added `InterfaceMode::Delayed` for the delayed behavior. I added some docstrings on `InterfaceMode` options. I figured the streaming mode is more similar to `cat` than `tail -f`. So I renamed the module to `cat.rs`.

markbt commented:
Thanks!  This looks good.

Naming things is hard.  What do you think about the name "direct" instead of "cat" to represent rendering directly to the terminal (rather than entering fullscreen mode)?

If you're happy with that, I'll do a search+replace and then merge it.

quark-zju commented:
Yeah, I'm not too happy with `cat`. `direct` looks good to me.

In some distince future I think it's more consistent for the "direct" interface to also support features like searching and showing line numbers. I didn't find an easy way to reuse `Screen` to do that, though.

markbt commented:
Cool.  Watching the video again, one idea might be for the non-alt-screen version to clear off the overlay lines at the bottom of the screen, so only the content is left.  That can happen later though, no need to do that now.  I'll see if I can find some time to do the search+replace tomorrow, and then merge these two requests.

Thanks again!

quark-zju commented:
I actually had the "clear overlay lines" implemented in a draft branch targeting at "theming" support (ex. options to behave like less - search and ruler uses a same line), where it switches to a "blank" theme to remove the overlays at exit. I found it difficult to get all line offset and refresh calculation done right cleanly, though. A React-like pattern might help, while preserving `ScrollRegion` fast paths can be challenging to implement cleanly.
